### PR TITLE
expand Boost branch

### PR DESCRIPTION
**Pull request by dkeeney**
This branch will install Boost and replace the preloaded include files in external/common/includes.  This first commit will not compile but deletes lots of files.

I am trying to make it easier to review by putting all of the deletions in a single commit.  The following commits will contain the changes to the CMake files to handle third party installation of Boost with the filesystem and system modules.

### CODE REVIEW

File: external/Boost.cmake [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+# You should have received a copy of the GNU Affero Public License
+# along with this program.  If not, see http://www.gnu.org/licenses.
+#
+# http://numenta.org/licenses/
+# -----------------------------------------------------------------------------
+
+# Creates ExternalProject for building the boost system and filesystem static libraries
+
+
+message(STATUS "----- Boost External Library ------")
```
Comment (breznak) [inline external/Boost.cmake]: :+1: for this external file

File: external/CMakeLists.txt [hunk 2]
```diff
@@ -19,14 +19,22 @@
...
 include(GNUInstallDirs)
 
+message(STATUS "----- external -----")
+message(STATUS "REPOSITORY_DIR = ${REPOSITORY_DIR}")
+message(STATUS "EP_BASE    = ${EP_BASE}")
 set_directory_properties(PROPERTIES EP_BASE "${EP_BASE}")
 
+# Determine common toolchain settings, compiler and link flags
+set(INTERNAL_CPP_STANDARD C++11)
+include("${REPOSITORY_DIR}/CommonCompilerConfig.cmake")
```
Comment (breznak) [inline external/CMakeLists.txt]: for external? I'd build external with default flags, and only use compiler confings for our sources?

File: src/CMakeLists.txt [hunk 3]
```diff
@@ -19,11 +19,18 @@
...
-cmake_minimum_required(VERSION 3.3)
+cmake_minimum_required(VERSION 3.7)
 project(nupic_core CXX)
 
+
+message(STATUS "------- nupic_core C++ library -------")
+message(STATUS "CMAKE_SYSTEM             = ${CMAKE_SYSTEM}")
+message(STATUS "CMAKE_BINARY_DIR         = ${CMAKE_BINARY_DIR}")
+message(STATUS "CMAKE_CURRENT_SOURCE_DIR = ${CMAKE_CURRENT_SOURCE_DIR}")
+message(STATUS "EP_BASE                  = ${EP_BASE}")
```
Comment (breznak) [inline src/CMakeLists.txt]: what is EP_BASE?

File: src/CMakeLists.txt [hunk 4]
```diff
@@ -40,6 +47,22 @@ if( POLICY CMP0054 )
   cmake_policy(SET CMP0054 OLD)
 endif()
 
+# Determine common toolchain settings, compiler and link flags
+include("${REPOSITORY_DIR}/CommonCompilerConfig.cmake")
+
+
+# To obtain values from the ExternalProjects for the dependencies
+# we load the cache from the 'external' ExternalProject.
+load_cache("${CMAKE_BINARY_DIR}/../external"
```
Comment (breznak) [inline src/CMakeLists.txt]: is this required? Could we use the common `include(external)` to build all those?

File: src/CMakeLists.txt [hunk 5]
```diff
@@ -303,91 +399,48 @@ set(src_nupiccore_srcs
...
 set(src_common_test_exe_libs
-    ${src_lib_static_nupiccore_combined}
+    ${src_lib_static_nupiccore}
     ${PYTHON_LIBRARIES}
     ${src_common_os_libs})
 
 message(STATUS "src_common_test_exe_libs = ${src_common_test_exe_libs}")
 
 
-## Example custom executable
```
Comment (breznak) [inline src/CMakeLists.txt]: let's keep one executable as an example

File: src/CMakeLists.txt [hunk 6]
```diff
@@ -397,12 +450,12 @@ add_executable(${src_executable_gtests}
                test/unit/algorithms/Cells4Test.cpp
                test/unit/algorithms/CondProbTableTest.cpp
                test/unit/algorithms/ConnectionsTest.cpp
-	       test/unit/algorithms/ConnectionsPerformanceTest.cpp
-	       test/unit/algorithms/HelloSPTPTest.cpp
+	      # test/unit/algorithms/ConnectionsPerformanceTest.cpp
+	      # test/unit/algorithms/HelloSPTPTest.cpp
```
Comment (breznak) [inline src/CMakeLists.txt]: re-enable these before merge

File: CMakeLists.txt [hunk 7]
```diff
@@ -70,19 +81,74 @@ if (MSYS OR MINGW)
...
+	CMAKE_ARGS -DCMAKE_BUILD_TYPE=${CMAKE_BUILD_TYPE}
+		-DREPOSITORY_DIR=${REPOSITORY_DIR}
+		-DNEEDS_BOOST=${NEEDS_BOOST}
+		-DEP_BASE=${EP_BASE}
+		-DNUPIC_BUILD_PYEXT_MODULES=${NUPIC_BUILD_PYEXT_MODULES}
+		-DPLATFORM=${PLATFORM}
+		-DBITNESS=${BITNESS}
+	BINARY_DIR ${CMAKE_BINARY_DIR}/external
+	STAMP_DIR  ${CMAKE_BINARY_DIR}
+	TMP_DIR /tmp/nupic
```
Comment (breznak) [inline CMakeLists.txt]: this won't work on Windows (but we dropped it for now)

File: CommonCompilerConfig.cmake [hunk 8]
```diff
@@ -110,14 +111,24 @@ set(EXTERNAL_LINKER_FLAGS_OPTIMIZED)
...
-if(CMAKE_SIZEOF_VOID_P EQUAL 8)
-  set(BITNESS 64)
+if(NOT BITNESS)
+  if(CMAKE_SIZEOF_VOID_P EQUAL 8)
+    set(BITNESS 64)
+  else()
+    set(BITNESS 32)
+  endif()
+endif()
+if(NOT INTERNAL_CPP_STANDARD)
```
Comment (breznak) [inline CommonCompilerConfig.cmake]: keep this on CMake, there's some CMake_CPP_version c++11)

File: CommonCompilerConfig.cmake [hunk 9]
```diff
@@ -309,8 +322,7 @@ endif()
 # Compatibility with gcc >= 4.9 which requires the use of gcc's own wrappers for
 # ar and ranlib in combination with LTO works also with LTO disabled
 IF(UNIX AND CMAKE_COMPILER_IS_GNUCXX AND (NOT "${CMAKE_BUILD_TYPE}" STREQUAL "Debug") AND
-      (CMAKE_CXX_COMPILER_VERSION VERSION_GREATER "4.9" OR
-       CMAKE_CXX_COMPILER_VERSION VERSION_EQUAL "4.9"))
+      (CMAKE_CXX_COMPILER_VERSION VERSION_GREATER_EQUAL "4.9"))
```
Comment (breznak) [inline CommonCompilerConfig.cmake]: drop this and require gcc-7+?

File: external/CMakeLists.txt (hunk 2 above)
Comment (dkeeney) [reply external/CMakeLists.txt]: I don't like calling CommonCompilerConfig.cmake twice (once here and once in src) but I wanted to compile the external stuff in C++11 and eventually src in C++17.  So the compile flags had to be setup differently.  Using default flags sounds good but default is different for different platforms. That is basically what CommonCompilerConfig figures out.

After we get rid of SWIG and APR we might be able to trim down CommonCompilerConfig to just the basics.

File: src/CMakeLists.txt (hunk 3 above)
Comment (dkeeney) [reply src/CMakeLists.txt]: Its where external stuff gets built.  build/scripts/ThirdParty

File: src/CMakeLists.txt (hunk 4 above)
Comment (dkeeney) [reply src/CMakeLists.txt]: What I wanted was a way to make the src stuff depend on all external dependencies.  The correct way to do that is with a superbuild where the src stuff is a separate ExternalProject which depends on the external ExternalProject.  
The problem with ExternalProjects is that you cannot pass arguments back up the hierarchy. The set( ... Parent Scope) does not work across ExternalProjects.  Using the cache does not work either unless you load the other ExternalProject's cache.

The correct way to export arguments out of ExternalProjects is to use find_package() at the places that you want to use the values.  And that is what I did for Boost and Yaml-cpp. But I cheated for the rest and use the cache.  Most of those eventually go away so we will end up with just the find_package() calls (if I can figure out a way to determine where Boost.cmake put BOOST_ROOT.)

File: src/CMakeLists.txt (hunk 5 above)
Comment (dkeeney) [reply src/CMakeLists.txt]: Hmmm, probably should, but they all went away.

File: src/CMakeLists.txt (hunk 6 above)
Comment (dkeeney) [reply src/CMakeLists.txt]: Of course, that is so I can build now.  will take them out after I merge in your changes.

File: CommonCompilerConfig.cmake (hunk 8 above)
Comment (dkeeney) [reply CommonCompilerConfig.cmake]: I actually let this be passed in from the SuperBuild.  The user can set this on the command line as an override.  Eventually  I will set this to C++17 as the default. but I have to get rid of some stuff first.

File: CommonCompilerConfig.cmake (hunk 9 above)
Comment (dkeeney) [reply CommonCompilerConfig.cmake]: Could I guess.

**Comment by dkeeney** [comment]
The way I envision BOOST being linked is the following.  Boost.cmake looks for boost:
- First we see if we even need Boost (will add this later when we add C++17 with std::filesystem)
- On the command line the user can specify BOOST_ROOT to override everything.
- An environment variable BOOST_ROOT.
- Standard places (i.e. /usr/local/Boost on Linux, etc.)
- If the correct flavor of boost is not found, download and build it.

File: src/CMakeLists.txt (hunk 4 above)
Comment (breznak) [reply src/CMakeLists.txt]: > a way to make the src stuff depend on all external dependencies. The correct way to do that is with a superbuild where the src stuff is a separate ExternalProject which depends on the external ExternalProject. 

That sounds reasonable, although the way seems little over-complicated to me, bt I'm no expert on cmake. Wouldn't `find_package(BOOST, REQUIRED)` suffice?

> Most of those eventually go away so we will end up with just the find_package() 

Then :+1:

**Comment by breznak** [comment]
> he way I envision BOOST being linked is the following. Boost.cmake looks for boost

Can/do we use this? 
https://cmake.org/cmake/help/v3.0/module/FindBoost.html

> My current problem is that unit_tests work for debug build but core dumps for the release build. 

Can you upload some +-working state? I'll try to dig in too.

Some idea for sub-goal, should we separate the tests to a stand-alone CMake and folder, and build them as "external" project/dependency? That could make this debug easier. 

Also, are you using same version of boost? Might be some lib version incompatibility. Should we update our deps?

**Comment by breznak** [comment]
@dkeeney 
Checked with the current branch, 
- detects system-wide boost on my system 
- linker error: 
```
[ 59%] Linking CXX static library libnupic_core_solo.a
[ 59%] Built target nupic_core_solo
[ 60%] Generating nupic_core_dummy.c++
Scanning dependencies of target nupic_core
[ 60%] Building CXX object CMakeFiles/nupic_core.dir/nupic_core_dummy.c++.o
[ 61%] Linking CXX static library libnupic_core.a
Combining $<TARGET_FILE:nupic_core> for target nupic_core from $<TARGET_FILE:yaml-cpp>;/usr/lib/x86_64-linux-gnu/libboost_system.so;/usr/lib/x86_64-linux-gnu/libboost_filesystem.so;/mnt/store/devel/HTM/htm-community/nupic.cpp/build/scripts/ThirdParty/Build/YamlStaticLib/libyaml.a;/mnt/store/devel/HTM/htm-community/nupic.cpp/build/scripts/ThirdParty/Install/Apr1StaticLib/lib/libapr-1.a;/mnt/store/devel/HTM/htm-community/nupic.cpp/build/scripts/ThirdParty/Install/AprUtil1StaticLib/lib/libaprutil-1.a;$<TARGET_FILE:nupic_core_solo>.
-- COMBINE_UNIX_ARCHIVES(  LIB_TARGET=nupic_core,   TARGET_LOCATION=/mnt/store/devel/HTM/htm-community/nupic.cpp/build/scripts/src/libnupic_core.a,   SRC_LIB_LOCATIONS=/mnt/store/devel/HTM/htm-community/nupic.cpp/build/scripts/ThirdParty/Build/YamlCppStaticLib/libyaml-cpp.a++++/usr/lib/x86_64-linux-gnu/libboost_system.so++++/usr/lib/x86_64-linux-gnu/libboost_filesystem.so++++/mnt/store/devel/HTM/htm-community/nupic.cpp/build/scripts/ThirdParty/Build/YamlStaticLib/libyaml.a++++/mnt/store/devel/HTM/htm-community/nupic.cpp/build/scripts/ThirdParty/Install/Apr1StaticLib/lib/libapr-1.a++++/mnt/store/devel/HTM/htm-community/nupic.cpp/build/scripts/ThirdParty/Install/AprUtil1StaticLib/lib/libaprutil-1.a++++/mnt/store/devel/HTM/htm-community/nupic.cpp/build/scripts/src/libnupic_core_solo.a,   LIST_SEPARATOR=++++,   BINARY_DIR=/mnt/store/devel/HTM/htm-community/nupic.cpp/build/scripts/src,   CMAKE_AR=gcc-ar)
-- COMBINE_UNIX_ARCHIVES: LIB_TARGET=nupic_core, src-lib=/mnt/store/devel/HTM/htm-community/nupic.cpp/build/scripts/ThirdParty/Build/YamlCppStaticLib/libyaml-cpp.a
-- COMBINE_UNIX_ARCHIVES: LIB_TARGET=nupic_core, src-lib=/usr/lib/x86_64-linux-gnu/libboost_system.so
/usr/bin/ar: /usr/lib/x86_64-linux-gnu/libboost_system.so: File format not recognized
CMake Error at /mnt/store/devel/HTM/htm-community/nupic.cpp/src/CombineUnixArchives.cmake:74 (message):
  COMBINE_UNIX_ARCHIVES: obj extraction process failed exe_result='1'
Call Stack (most recent call first):
  /mnt/store/devel/HTM/htm-community/nupic.cpp/src/CombineUnixArchives.cmake:124 (combine_unix_archives)


CMakeFiles/nupic_core.dir/build.make:104: recipe for target 'libnupic_core.a' failed
make[5]: *** [libnupic_core.a] Error 1
CMakeFiles/Makefile2:354: recipe for target 'CMakeFiles/nupic_core.dir/all' failed
make[4]: *** [CMakeFiles/nupic_core.dir/all] Error 2
make[4]: *** Waiting for unfinished jobs....
Prepending /mnt/store/devel/HTM/htm-community/nupic.cpp/src/nupic/bindings/swig_proxy_preamble.py to swig-generated module /mnt/store/devel/HTM/htm-community/nupic.cpp/build/scripts/src/math.py
[ 61%] Built target _math
Makefile:151: recipe for target 'all' failed
make[3]: *** [all] Error 2
```

Esp. 
> /usr/bin/ar: /usr/lib/x86_64-linux-gnu/libboost_system.so: File format not recognized
CMake Error at /mnt/store/devel/HTM/htm-community/nupic.cpp/src/CombineUnixArchives.cmake:74 (message):
  COMBINE_UNIX_ARCHIVES: obj extraction process failed exe_result='1' 

I think there's a problem with using the binary boost I have on system (why?) So we must force source build (?) 

- Also, about the Debug ok and Release failing, try to disable the `-flto` and other optimizations temporarily.

**Comment by breznak** [comment]
- if I uninstall system boost (1.65), it compiles and runs even for Release and unit-tests. 
- py tests fail 
> _________________________________________ ERROR collecting bindings/py/tests/array_algorithms_test.py ________________________________________________
ImportError while importing test module '/mnt/store/devel/HTM/htm-community/nupic.cpp/bindings/py/tests/array_algorithms_test.py'.
Hint: make sure your test modules/packages have valid Python names.                                                                                          
Traceback:                                                                                                                                                   
array_algorithms_test.py:29: in <module>                                                                                                                     
    from nupic.bindings.math import nearlyZeroRange                                                                                                          
../src/nupic/bindings/math.py:70: in <module>                                                                                                                
    _math = swig_import_helper()                                                                                                                             
../src/nupic/bindings/math.py:66: in swig_import_helper                                                                                                      
    _mod = imp.load_module('_math', fp, pathname, description)                                                                                               
E   ImportError: /mnt/store/devel/HTM/htm-community/nupic.cpp/bindings/py/src/nupic/bindings/_math.so: undefined symbol: _ZTIN5nupic10NumpyArrayE    

Seems a problem with Numpy includes, can you try to minimize even more the cmake changes? How important is the super-build paradigm? If we build boost as all other deps ... 

- (very minor) build option `make tests_all` disappeared, which I used to use, but guess it's ok to rm.

**Comment by dkeeney** [comment]
The way it is now, we can ONLY use a static Boost library.  We cannot use .so  So I was expecting my find_package(Boost ) to reject the dynamic library.  The reason is that we merge the libraries, including the Boost library, together into a single nupic_core static library.  Why?  Don't know.   The original CMake combined all external libraries into a single static library and I just assumed there was some reason they did that.

Now, maybe we should use the dynamic libraries instead.  That would simplify a lot of things.    But would it run in the Python 2.7 environment?   

Something to think about.  Thanks.

**Comment by breznak** [comment]
> ... into a single nupic_core static library. Why? Don't know. The original CMake combined all external libraries into a single static library and I just assumed there was some reason they did that. 

Try to look here, 
https://github.com/numenta/nupic.core/issues?utf8=%E2%9C%93&q=is%3Aissue+is%3Aclosed+static+lib

Might have been capnp, pybindings, or (my tip) that they wanted binary bindings to work on all systems (we can drop that requirement and assume source build always) . 

> Now, maybe we should use the dynamic libraries instead. That would simplify a lot of things. But would it run in the Python 2.7 environment? 

I would be for this option, if applicable for py2.7/swig (they'll go away anyway)

**Comment by dkeeney** [comment]
Well, what I am doing is creating a new local branch from master and try getting the SWIG/Python to work with a dynamically linked Boost.  If that does not work we could still require Boost to be installed externally (not part of our package) and link statically to it.

I just now manually installed the latest Boost into /usr/local/lib and it installed both dynamic and static libraries. So even if I cannot get dynamic lib to work we can do this without the Boost.cmake or the Super Build.

That would be much simpler.

**Comment by breznak** [comment]
@dkeeney is there anything else you wanted to extract from this branch, or we can close it as obsoleted by the just-merged #102 ?

**Pull request closed without merging**